libff/algebra/curves/profile/algebra_groups_read_profile.cpp

```diff
@@ -75,40 +90,647 @@ bool profile_group_read_sequential_uncompressed(const std::string &identifier)
+            nullptr,
+            file_size,
+            PROT_READ,
+            MAP_PRIVATE /* | MAP_NOCACHE */,
+            fd,
+            0);
+        if (MAP_FAILED == file_base) {
+            throw std::runtime_error(
+                std::string("mmap failed: ") + strerror(errno));
+        }
+
+        const size_t element_interval_bytes = element_size_on_disk * interval;
+
+        {
+            enter_block("Read group elements profiling");
+            for (size_t i = 0; i < NUM_ELEMENTS_TO_READ; ++i) {
+                const size_t different_element_idx = i % NUM_DIFFERENT_ELEMENTS;
+                const size_t offset = i * element_interval_bytes;
+                GroupT &dest_element = elements[different_element_idx];
+                const GroupT *src =
+                    (const GroupT *)(const void *)((size_t)file_base + offset);
+
+                dest_element.X = src->X;
+                dest_element.Y = src->Y;
+            }
+            leave_block("Read group elements profiling");
+        }
+
+        if (0 != munmap((void *)file_base, file_size)) {
+            throw std::runtime_error("munmap failed");
+        }
+        close(fd);
+    }
+}
+
+void cb_init(
+    aiocb *cb, int fd, size_t offset_bytes, size_t size_bytes, void *dest)
+{
+    // std::cout << "cb_init: " << fd << ", off: " << offset_bytes << ", size: "
+    // << size_bytes << ", dst: " << dest << "\n";
+
+    memset(cb, 0, sizeof(aiocb));
+    cb->aio_fildes = fd;
+    cb->aio_offset = offset_bytes;
+    cb->aio_buf = dest;
+    cb->aio_nbytes = size_bytes;
+    cb->aio_sigevent.sigev_notify = SIGEV_NONE;
+    cb->aio_lio_opcode = LIO_READ;
+}
+
+void cb_enqueue(aiocb *cb)
+{
+    const int r = aio_read(cb);
+    if (0 == r) {
+        // std::cout << "cb_wait: enqueued\n";
+        return;
+    }
+
+    throw std::runtime_error(
+        std::string("error from aio_read: ") + strerror(errno));
+}
+
+ssize_t cb_wait(aiocb *cb)
+{
+    int err;
+    for (;;) {
+        err = aio_error(cb);
+        if (EINPROGRESS == err) {
+            std::this_thread::yield();
+            // sleep(0);
+            continue;
+        }
+
+        const ssize_t ret = aio_return(cb);
+        if (0 <= ret) {
+            // std::cout << "cb_wait: done\n";
+            return ret;
+        }
+
+        if (ECANCELED == err) {
+            throw std::runtime_error("aio_error: cancelled");
+        }
+
+        throw std::runtime_error(
+            std::string("error from aio_error: ") + strerror(errno));
+    }
+}
+
+template<typename GroupT>
+void profile_group_read_random_aio_ordered_uncompressedw(
+    const std::string &identifier, const size_t interval)
+{
+    const std::string filename = get_filename(identifier);
+
+    // Treat the file as being divided into NUM_ELEMENTS_TO_READ groups, of
+    // size NUM_ELEMENTS_IN_FILE / NUM_ELEMENTS_TO_READ, reading from a ranadom
+    // offset within each group.
+
+    const size_t GROUP_SIZE = interval;
```

Comment:
What about renaming "group" to "batch" here? i.e. `Treat the file as being divided into NUM_ELEMENTS_TO_READ groups` -> `Treat the file as being divided into NUM_ELEMENTS_TO_READ batches` and do the relevant changes in the code.

Here we process group elements in group. I find the wording slightly confusing, and think that would go away by simply speaking of "batches" and/or "chunks". That's not a very big deal, reading the code makes the situation clearer, but I just happened to lend on this file and read "GROUP_SIZE" as if it was the number of elements in the (algebraic group), and then realized, "Ah! no, ok". Maybe we can avoid unnecessary word collisions here. Unless I am misunderstanding something.

Reply:
I see.  No, it's certainly not the number of elements in the group - that's not of concern here.  "Batch" may be a little bit misleading, since it sounds like a set of *oeprations*.  I've removed GROUP_SIZE and just used `interval` everywhere (since this value is the average interval between read locations in the file).

Reply:
>  just used interval everywhere

Actually, what about renaming `interval` too here? To me, an interval is a **set** of numbers lying between 2 numbers of the set (e.g. the real interval [1, 10] is the set of numbers between 1 and 10 - bounds included). However, in the code, `interval` represents a "chunk size" or "interval size" used to consume the files. I think this is slightly misleading to have a notion referring to a "set" to be of type "const size_t" and used as if it referred to a "set cardinality" (e.g. `const size_t element_interval_bytes = element_size_on_disk * interval` is slightly odd to read as you wonder what operation `*` does on a set, and then realize by referring to the type that, ok, it's the usual "multiplication" in the ints)

Reply:
"Interval" can also mean the gap between things (perhaps the more common usage) rather than a range.  I've updated the comment to explicitly mention the "interval betwen reads", so hopefully that should eliminate any ambiguity.  I think "interval" is the most natural way to refer to this.

If you still find it confusing, I'd rather switch to something like `read_sparsity`.  This interval between reads is the important concept in this function, rather than virtual "sections" or "chunks" of the file.

Reply:
I've adjusted the comments again to try and clarify the point here.  Hopefully that explains the point about `interval` fully.

Reply:
Thanks